src/FindIntrinsics.cpp

```diff
@@ -575,6 +583,57 @@ class FindIntrinsics : public IRMutator {
             return rewrite.result;
         }
 
+        const int bits = op->type.bits();
+        const auto is_x_same_int = op->type.is_int() && is_int(x, bits);
+        const auto is_x_same_uint = op->type.is_uint() && is_uint(x, bits);
+        const auto is_x_same_int_or_uint = is_x_same_int || is_x_same_uint;
+        auto x_y_same_sign = (is_int(x) == is_int(y)) || (is_uint(x) && is_uint(y));
+        Type unsigned_type = op->type.with_code(halide_type_uint);
+        const auto is_x_wider_int_or_uint = (op->type.is_int() && is_int(x, 2 * bits)) || (op->type.is_uint() && is_uint(x, 2 * bits));
+        Type opposite_type = op->type.is_int() ? op->type.with_code(halide_type_uint) : op->type.with_code(halide_type_int);
+        const auto is_x_wider_opposite_int = (op->type.is_int() && is_uint(x, 2 * bits)) || (op->type.is_uint() && is_int(x, 2 * bits));
+
+        if (
+            // Saturating patterns.
+            rewrite(saturating_cast(op->type, widening_add(x, y)),
+                    saturating_add(x, y),
+                    is_x_same_int_or_uint) ||
+            rewrite(saturating_cast(op->type, widening_sub(x, y)),
+                    saturating_sub(x, y),
+                    is_x_same_int_or_uint) ||
+            rewrite(saturating_cast(op->type, shift_right(widening_mul(x, y), z)),
+                    mul_shift_right(x, y, cast(unsigned_type, z)),
```

Comment:
Why is there a cast to the unsigned type of z if the predicate says it's already an unsigned type?

Reply:
I think a combination of copy-paste and a lack of double-checking. I will fix it

Reply:
Oh, no. The cast is actually a narrowing cast, because `unsigned_type` is the output type bitwidth. This matches the original patterns here: https://github.com/halide/Halide/blob/9ca75606a0a9c8fa43c1dee82db734cc3d0f7e39/src/FindIntrinsics.cpp#L454-#L469
I am now convinced that this is actually a bug, we should be checking if `z` can be losslessly casted to the unsigned type.
On current main, I see this buggy behavior:
```
Type t = Int(16, 16);
Expr x = Variable::make(t, "x");
Expr y = Variable::make(t, "y");
Expr z = Variable::make(UInt(32, 16), "z");
Expr expr = cast(t, clamp(widening_mul(x, y) >> z, t.min(), t.max()));
std::cout << expr << "\n";
std::cout << find_intrinsics(expr) << "\n";
```
```
int16x16(max(min((int32x16)shift_right((int32x16)widening_mul((int16x16)x, (int16x16)y), (uint32x16)z), x16(32767)), x16(-32768)))
(int16x16)mul_shift_right((int16x16)x, (int16x16)y, uint16x16((uint32x16)z))
```
I believe this is a bug, depending on our definition of `shift_right`. Is it defined for all values of the rhs or only `[-bits, +bits]`?

Reply:
If the latter is true, then this is a safe lifting rule to make, otherwise we may need to fix multiple rules

Reply:
Discussed with @abadams  offline, the assumption here is that we assume no out-of-range shifts can be present in the input, therefore these rules are perfectly valid.